.github/workflows/ci.yml

```diff
@@ -136,14 +136,94 @@ jobs:
         working-directory: crab-boil
         run: |
           ./criticalup link remove
-  
+
       - name: Run test cleanup workflow
         working-directory: crab-boil
         run: |
           ./criticalup remove
           ./criticalup clean
 
-          
+  integration-test-file-server:
+    name: Test CriticalUp-Dev against local file server on GHA Runners
+    strategy:
+      fail-fast: false
+    runs-on: ubuntu-latest
+    env:
+      CRITICALUP_DEV_TOKEN: ${{ secrets.CRITICALUP_DEV_TOKEN }}
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: "Define toolchain"
+        shell: bash # Powershell does not do heredocs
+        run: |
+          mkdir -p crab-boil
+          cd crab-boil
+          cat <<- EOF > criticalup.toml
+            manifest-version = 1
+
+            [products.ferrocene]
+            release = "stable-25.02.0"
+            packages = [
+                "cargo-\${rustc-host}",
+                "rustc-\${rustc-host}",
+                "rust-std-\${rustc-host}",
+            ]
+          EOF
+
+      - uses: actions/download-artifact@v4
+        with:
+          name: file-server
+          path: crab-boil
+          github-token: ${{ secrets.TEMPORARY_ACCESS_TOKEN_TO_AIR_GAPPED_ACTIONS_ARTIFACT }}
+          repository: ferrocene/criticalup-air-gapped-server
+          run-id: 16942640950
+
+      - name: make criticalup-dev binary
+        working-directory: crab-boil
+        run: |
+          CARGO_TARGET_DIR=target cargo build --release -p criticalup-dev
+          cp target/release/criticalup-dev ./criticalup-dev
+          chmod +x ./criticalup-dev
+
+      - name: Auth Criticalup
+        working-directory: crab-boil
+        run: ./criticalup-dev auth set $CRITICALUP_DEV_TOKEN
+
+      - name: Install toolchain from dev server, in order to get the keys and artifacts
+        working-directory: crab-boil
+        run: |
+          ./criticalup-dev install
+
+      - name: Copy artifacts and keys.json to local server. Run local server
+        working-directory: crab-boil
+        run: |
+          mkdir -p ./artifacts/products/ferrocene/releases/stable-25.02.0
+          cp  ~/.cache/criticalup-dev/artifacts/products/ferrocene/releases/stable-25.02.0/* ./artifacts/products/ferrocene/releases/stable-25.02.0/
+          cp ~/.cache/criticalup-dev/keys.json ./
+          chmod +x ./criticalup-air-gapped-server
+          ./criticalup-air-gapped-server serve &
+
+      - name: Remove toolchain and cache generated by using remote dev download server
+        working-directory: crab-boil
+        run: |
+          ./criticalup-dev remove
+          ./criticalup-dev clean
+
+
+      - name: Install toolchain from local server
+        working-directory: crab-boil
+        run: |
+          DOWNLOAD_SERVER_URL=http://localhost:3000 ./criticalup-dev install
+
+      - name: Run `criticalup-dev` with artifacts from local server in test workflow
```

Comment:
I don't love that it feels like there is a divergence risk between this job and the other test job, but I think it's fine as a first step.